### DISCUSSION TITLE
Where jitterbuffer should be used?

### DISCUSSION THREAD

**Discussion by vipcxj** [pion/webrtc#2811] [Q&A]
I use webrtc in a cluster(multi server).
client A <-webrtc-> server A <- udp -> server B <-webrtc-> client B
client A communicate with  server A using webrtc; server A send the rtp packet to server B using udp; server B read the rtp packet from udp and write it to a TrackLocalStaticRTP which is the track of a peer to client B.
Currently it works. However, even in a reliable environment such as the Intranet,client B still encountered a large number of errors when parsing video streams.
My question is should I add a jitter buffer after read the rtp package from the udp? Of course, There already has been jitter buffer in client B when accept packet from the server B.

**Comment by diegomoreira001** [comment]
There a re some use-cases where you can skip it and you may get away with it.

1. Over a simple LAN, jittering is almost non-existent (take it with a big grain of salt, im assuming the network topology is fairly simple).
2. Over a LAN and Internet, if you built an SFU that only needs to forward packets and do nothing else, you may relay on the client's browser buffering to solve jittering for you. However, it would be advisable to implement a NACK interceptor on the SFU. Otherwise the client's jitter buffer will kinda work but wont receive any lost packets. 

So, while jitter may be an issue, NACK and PLI are always more important, since you need to detect dropped packets and dropped frames. 

I'd only implement a jitter buffer if your SFU needs to do some processing on its own, or the destination is not a full-fledged browser (e.g. saving to disk the track, streaming the track to a service such as voice recognition, etc).

Anyway, that's my take on this. There are more experienced people in here that can correct if im wrong.

Cheers!

**Comment by vipcxj** [comment]
Hi @diegomoreira001 , I log every packets written to TrackLocalStaticRTP with its seq number and md5 (from pos 12 byte, because I think some head bit is changed). And I log every rtp packet received on server B with its seq number and md5 (from pos 12 byte) as well. But they are not equal (And I can't find any md5 which exist in TrackLocalStaticRTP side exist in Server B side as well). Will TrackLocalStaticRTP repacket the input stream base on the wrong order when not using jitterbuffer? Even I test it in a local network, I found many lost packets on server B
